File: app/services/llm_service.py
import json
from typing import Any, Dict
from openai import AsyncOpenAI
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type
from app.core.config import get_settings
from app.services.prompt_builder import SYSTEM_PROMPT, build_user_prompt


class LLMService:
    def __init__(self) -> None:
        settings = get_settings()
        self.client = AsyncOpenAI(api_key=settings.openai_api_key, timeout=settings.openai_timeout_seconds)
        self.model = settings.openai_model

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=0.2, min=0.2, max=1.5),
        retry=retry_if_exception_type(Exception),
        reraise=True,
    )
    async def analyze(self, context: Dict[str, Any]) -> Dict[str, Any]:
        response = await self.client.chat.completions.create(
            model=self.model,
            temperature=0.1,
            response_format={"type": "json_object"},
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": build_user_prompt(context)},
            ],
        )
        content = response.choices[0].message.content or "{}"
        return json.loads(content)

File: src/services/llmService.js
const { createConfiguredLlmProvider } = require("./llmProviders");

class LLMService {
  constructor({ provider = createConfiguredLlmProvider() } = {}) {
    this.provider = provider;
  }

  getMetadata() {
    return this.provider.getMetadata
      ? this.provider.getMetadata()
      : { provider: "unknown", model: "unknown" };
  }

  async analyze(context) {
    return this.provider.analyze(context);
  }
}

module.exports = { LLMService };
